File: src/main/java/org/openhds/mobile/search/SimpleCursorDocumentSource.java
package org.openhds.mobile.search;

import android.database.Cursor;

import org.apache.lucene.document.Field;

class SimpleCursorDocumentSource extends CursorDocumentSource {

    private static final int UUID_COL_IDX = 0;
    private static final int LEVEL_COL_IDX = 1;

    SimpleCursorDocumentSource(Cursor c) {
        super(c);
    }

    @Override
    public Field[] getFields() {
        int columns = cursor.getColumnCount();
        Field[] fields = new Field[columns];
        for (int c = 0; c < columns; c++) {
            fields[c] = new Field(getFieldName(c), "", getFieldStore(c), getFieldIndex(c));
        }
        return fields;
    }

    public String getFieldName(int index) {
        return cursor.getColumnName(index);
    }

    public Field.Store getFieldStore(int index) {
        return index == UUID_COL_IDX || index == LEVEL_COL_IDX ? Field.Store.YES : Field.Store.NO;
    }

    public Field.Index getFieldIndex(int index) {
        if (index == UUID_COL_IDX) {
            return Field.Index.NO;
        } else if (index == LEVEL_COL_IDX) {
            return Field.Index.NOT_ANALYZED;
        } else {
            return Field.Index.ANALYZED;
        }
    }

    @Override
    public String getFieldValue(int index) {
        return cursor.getString(index);
    }
}

File: src/main/kotlin/org/cimsbioko/search/SimpleCursorDocumentSource.kt
package org.cimsbioko.search

import android.database.Cursor
import org.apache.lucene.document.Field
import org.apache.lucene.document.Field.Index
import org.apache.lucene.document.Field.Store

internal open class SimpleCursorDocumentSource(c: Cursor) : CursorDocumentSource(c) {

    override val fields: Array<Field>
        get() = with(mutableListOf<Field>()) {
            fun getFieldStore(index: Int) = if (index == UUID_COL_IDX || index == LEVEL_COL_IDX) Store.YES else Store.NO
            fun getFieldIndex(index: Int) = when (index) {
                UUID_COL_IDX -> Index.NO
                LEVEL_COL_IDX -> Index.NOT_ANALYZED
                else -> Index.ANALYZED
            }
            for (c in 0 until cursor.columnCount) {
                add(Field(getFieldName(c), "", getFieldStore(c), getFieldIndex(c)))
            }
            toTypedArray()
        }

    fun getFieldName(index: Int): String {
        return cursor.getColumnName(index)
    }

    override fun getFieldValue(index: Int): String? {
        return cursor.getString(index)
    }

    companion object {
        private const val UUID_COL_IDX = 0
        private const val LEVEL_COL_IDX = 1
    }
}